oap-server/server-storage-plugin/storage-elasticsearch-plugin/src/main/java/org/apache/skywalking/oap/server/storage/plugin/elasticsearch/base/MetricsEsDAO.java

```diff
@@ -117,4 +121,25 @@ public UpdateRequest prepareBatchUpdate(Model model, Metrics metrics) throws IOE
         String id = IndexController.INSTANCE.generateDocId(model, metrics.id());
         return getClient().prepareUpdate(modelName, id, builder);
     }
+
+    @Override
+    public boolean isExpiredCache(final Model model,
+                                  final Metrics cachedValue,
+                                  final long currentTimeMillis,
+                                  final int ttl) {
+        final long metricTimestamp = TimeBucket.getTimestamp(
+            cachedValue.getTimeBucket(), model.getDownsampling());
+        // Fast fail check. If the duration is still less than TTL - 1 days(absolute)
+        // the cache should not be expired.
+        if (currentTimeMillis - metricTimestamp < TimeUnit.DAYS.toMillis(ttl - 1)) {
+            return false;
+        }
+        final long deadline = Long.parseLong(new DateTime(currentTimeMillis).plusDays(-ttl).toString("yyyyMMdd"));
+        final long timeBucket = TimeBucket.getTimeBucket(cachedValue.getTimeBucket(), DownSampling.Day);
+        // If time bucket is earlier or equals(mostly) the deadline, then the cached metric is expired.
+        if (timeBucket <= deadline) {
```

Comment:
This keeps consistent with TTL mechanism implementation.

https://github.com/apache/skywalking/blob/4003c4b777d8e4026e87a4474ede939c15282b05/oap-server/server-storage-plugin/storage-elasticsearch-plugin/src/main/java/org/apache/skywalking/oap/server/storage/plugin/elasticsearch/base/HistoryDeleteEsDAO.java#L62-L63